Index: linux-5.15.162/drivers/net/phy/realtek.c
===================================================================
--- linux-5.15.162.orig/drivers/net/phy/realtek.c
+++ linux-5.15.162/drivers/net/phy/realtek.c
@@ -968,6 +968,10 @@ static int rtl8221b_config_init(struct p
 	phy_read_mmd_poll_timeout(phydev, RTL8221B_MMD_SERDES_CTRL, 0x7587,
 				  val, !(val & BIT(0)), 500, 100000, false);

+	/* LED */
+	phy_write_mmd(phydev, RTL8221B_MMD_PHY_CTRL, 0xD034, 0x20);
+	phy_write_mmd(phydev, RTL8221B_MMD_PHY_CTRL, 0xD036, 0x07);
+
 	return 0;
 }
